Shipping rates UI changes: Make Edit Free Listings support country-level free shipping and minimum order amount

ecgan opened the pull request:
### Changes proposed in this Pull Request:

- Part of a PR refactor from the huge PR https://github.com/woocommerce/google-listings-and-ads/pull/1218.
- Depends on PR:
  - https://github.com/woocommerce/google-listings-and-ads/pull/1262
    - https://github.com/woocommerce/google-listings-and-ads/pull/1202
    - https://github.com/woocommerce/google-listings-and-ads/pull/1249
- Partly addresses the tech debt: https://github.com/woocommerce/google-listings-and-ads/issues/288.

This PR contains code refactoring on the shipping rate section and make the Edit Free Listings support country-level free shipping and minimum order amount setup.

#### Prior to this PR

In **Edit Free Listings**, the shipping rate code has the following file structure:

![image](https://user-images.githubusercontent.com/417342/154834431-434c7d3b-0a4a-4dbd-9094-09e987473430.png)

They are mostly "dumb components" with no API calls and no knowledge of wp.data store. Values are stored in the parent Form component, and there is no auto-save feature like in Setup MC.

#### In this PR

Shipping rate section is now a shared component called `ShippingRateSection`. The following is the files structure:

![image](https://user-images.githubusercontent.com/417342/154722859-d149cdd0-a19c-4421-9db5-3310408b6c97.png)

Here's a visualization of the component composition structure: 

![image](https://user-images.githubusercontent.com/417342/154724830-1abdd7e7-7d21-4ec3-a6fc-f7f39fef99f6.png)

All the components highlighted above are dumb components; they do not have any API calls in them. Shipping rates data are retrieved and stored in the parent form state (with the form key `shipping_country_rates`), and are propagated up and down to the components via `value` and `onChange` props.

`FlatShippingRatesInputCards` will only show up when users select flat shipping rate option. It contains `EstimatedShippingRatesCard` and `FreeShippingCards`. 

`EstimatedShippingRatesCard` is from [`ShippingCountriesForm`](https://github.com/woocommerce/google-listings-and-ads/blob/a6963bc069d58bb0ebf361ae41f9a1633de0b4da/js/src/components/free-listings/configure-product-listings/shipping-rate/shipping-rate-setup/countries-form.js), refactored and simplified. The values in this card depends on the `shipping_country_rates` form value.

`FreeShippingCards` is a wrapper around `OfferFreeShippingCard` and `MinimumOrderCard`. It will only be displayed when there is a shipping rate bigger than 0. It contains `useState` with _initial value_ depending on the `shipping_country_rates` form value; subsequent value depends on users selection of "Yes" / "No". When "Yes" is selected, it will render `MinimumOrderCard`.

`MinimumOrderCard` is a new component. It depends on the `shipping_country_rates` form value.

### Screenshots:

Shipping Rates section, with the new "offer free shipping" card and "minimum order" card. This UI and the code is the same for both Setup MC flow and Edit Free Listings flow.

![image](https://user-images.githubusercontent.com/417342/154556792-942a6d6e-9254-4bb6-8c3c-6662a956672a.png)

📹  Demo video with my voice annotation:

https://user-images.githubusercontent.com/417342/154854800-d3d9a1a7-d93e-4ebf-b807-520e4ea5fb11.mov


### Detailed test instructions:

Note: 

1. Open the browser dev tools network panel to observe the API calls.
2. You might want to test the Edit Free Listings first, it may be simpler to understand and see the interactions between the components, without needing to care about the auto-save API calls.

#### Edit Free Listings

Make sure the Setup MC is already completed, otherwise you won't be able to view the Edit Free Listings page. You may complete Setup MC using develop branch, and then switch to this PR branch to test this Edit Free Listings.

1. Go to `/wp-admin/admin.php?page=wc-admin&path=%2Fgoogle%2Fdashboard&subpath=%2Ffree-listings%2Fedit&pageStep=2`. You will see Edit Free Listings - Step 2.
2. When "flat" shipping rate is selected, you should see the "Estimated shipping rates" card.
3. Set some shipping rates to 0 value. You should see the green pill with "Free shipping for all orders" text.
3. Set some shipping rates to be bigger than 0. You should **not** see the green pill with "Free shipping for all orders" text.
4. When there is a non-zero shipping rate, the "offer free shipping" card should appear. "Yes" should be selected when there is an existing "minimum order amount", "No" otherwise.
5. In "offer free shipping" card, when "Yes" is selected, the "minimum order" card should appear. When "No" is selected, it would remove all the "minimum order amount" values that you have entered, and the "minimum order" card would disappear. When "Yes" is re-selected, the "minimum order" card would appear, with one visible input without any value.
6. In the minimum order card, the label should display at most 5 country names, with a suffix "+ <number> more" indicating the number of more countries.
7. Edit the values in the minimum order card, via inline edit and the edit modal (by clicking on the Edit button). It should work as expected.
7. In minimum order card, "Add another minimum order" button should appear when you have a country with non-zero shipping rate and does not have a minimum order defined yet. Clicking on the button should display a "add minimum order" modal, and the country should be pre-selected in the country SelectControl. 
8. Removing a country from shipping rates card should remove the country from minimum order card.
9. Click on the "Save changes" button in the page. You should see a notice that says saving is successful. Reload the page and make sure that the values are same as before the reload.
10. Repeat the above steps a few more times trying out different value permutations. 

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with `(Fix|Add|Tweak|Update) - `, for example:
> Fix - I took care of something that wasn't working.
> Add - I added something new that's pretty cool.
> Tweak - I made a small change.
> Update - I made big changes to something that wasn't broken.

Leave the "Changelog entry" header in place completely empty, without any summary if no changelog entry is needed.
If you remove the "Changelog entry" header, the title of Pull Request will be used as the changelog entry.
-->
### Changelog entry

>

ecgan committed:
Merge branch 'feature/shipping-rates-automatic-settings-ui' into feature/shipping-rates-api-ui-changes

ecgan committed:
Accomodate changes to shipping rates suggestions API.

ecgan committed:
Return shipping rates suggestions data as is.

The suggestions data schema is the same as shipping rates schema. We don't need any transformation.

ecgan committed:
Accommodate changes to saving shipping rate suggestions API.

ecgan committed:
Accommodate changes to fetchShippingRates response data.

The response data is now an array of shipping rates. Each shipping rate has its own id.

ecgan committed:
Merge branch 'feature/modify-shipping-rate-api' into feature/shipping-rates-api-ui-changes

Conflicts:
	js/src/setup-mc/setup-stepper/setup-free-listings/shipping-rate/shipping-rate-setup/useSaveSuggestions.js

ecgan committed:
Make groupShippingRatesByPriceCurrency accommodate shipping rate schema change.

shippingRate.countryCode has been changed to shippingRate.country.

ecgan committed:
Fix upsertShippingRates in wp-data.

shippingRates is now an array of shippingRate, not aggregated.

ecgan committed:
Fix null check in UPSERT_SHIPPING_RATES.

ecgan committed:
Refetch shipping rates after upsert.

We need this because some of the shipping rates may be newly created and do not have the id, so we refetch and store them in wp-data.

ecgan committed:
wp-data changes for deleteShippingRates.

ecgan committed:
Change Setup MC AddRateModal and EditRateModal to work with upsertShippingRates and deleteShippingRates.

ecgan committed:
Merge branch 'feature/refactor-shipping-rates-api' into feature/shipping-rates-api-ui-changes

ecgan committed:
Merge branch 'feature/refactor-shipping-rates-api' into feature/shipping-rates-api-ui-changes

ecgan committed:
Remove offers_free_shipping and free_shipping_threshold settings from front-end.

ecgan committed:
Change countryCode to country in ShippingCountriesForm.

This is to make the new shippingRate schema work in Edit Free Listings.

ecgan committed:
Refactor ShippingRateSetup.

ShippingRateSetup doesn't need formProps, just need value and onChange, and it does not need to know the field name 'shipping_country_rates'.

ecgan committed:
Move and refactor ShippingRateSetup with FlatShippingRatesInput.

ecgan committed:
Rename FlatShippingRatesInput to FlatShippingRatesInputCards.

ecgan committed:
Remove unneeded EstimatedShippingRatesCard folder.

ecgan committed:
Make upsertShippingRates work for Edit Free Listings.

ecgan committed:
Display green free shipping pill when shipping rate is 0.

ecgan committed:
Make free shipping pill wrap with better layout.

ecgan committed:
Code restructure.

ecgan committed:
Convert shipping rate from string to number.

ecgan committed:
Fix shipping rate options should be object, not array.

ecgan committed:
Display OfferFreeShippingCard.

ecgan committed:
Display MinimumOrderCard.

ecgan committed:
Fix free_shipping_threshold typo.

ecgan committed:
Remove free_shipping_threshold when user selects 'no' for free shipping.

ecgan committed:
Display MinimumOrderCard with MinimumOrderInput.

ecgan committed:
Consolidate 'yes' and 'no' values within OfferFreeShippingCard.

ecgan committed:
Code refactor in MinimumOrderCard.

ecgan committed:
Label and layout for MinimumOrderInput.

ecgan committed:
Code refactor with MinimumOrderInputLabel.

ecgan committed:
Add edit minimum order button and modal.

ecgan committed:
Rename MinimumOrderInput to MinimumOrderInputControl.

ecgan committed:
Rename MinimumOrderInputLabel to MinimumOrderInputLabelText.

ecgan committed:
Make edit work in EditMinimumOrderModal.

ecgan committed:
Make delete work in EditMinimumOrderModal.

ecgan committed:
Fix NaN threshold by checking free_shipping_threshold before calling Number().

ecgan committed:
Set threshold to undefined to prevent 0.

ecgan committed:
Code restructure with FreeShippingCards.

This is so that when users remove all the shipping rates, the OfferFreeShippingCard and MinimumOrderCard will not be rendered.

ecgan committed:
Add code for renderGroups in MinimumOrderCard.

ecgan committed:
Code refactor for renderGroups in MinimumOrderCard.

ecgan committed:
Allow add minimum order.

ecgan committed:
Fix filtering shipping rates in MinimumOrderCard.

ecgan committed:
Fix edit handler in MinimumOrderCard.

ecgan committed:
Better card title layout.

ecgan committed:
Make WCDL SelectControl label look like other input label.

ecgan committed:
Simplify SelectControl label in EditRateModal.

ecgan committed:
Simplify layout code for add and edit rate modals.

ecgan committed:
Move AddRateModal and EditRateModal files.

ecgan committed:
Move files into minimum-order-card folder.

ecgan committed:
Replace AddRateButton with AppButtonModalTrigger.

ecgan committed:
Replace EditRateButton with AppButtonModalTrigger.

ecgan committed:
Refactor groupShippingRatesByFreeShippingThreshold to separate file.

ecgan committed:
Remove unneeded audienceCountries check and AppSpinner render.

ecgan committed:
Rename CountriesPriceInput to ShippingRateInputControl.

ecgan committed:
Make ShippingCountriesForm render card on its own and remove unnecessary ShippingRateSetup intermediate component.

ecgan committed:
Rename ShippingCountriesForm to EstimatedShippingRatesCard, and move files.

ecgan committed:
Fix export error for MinimumOrderCard.

ecgan committed:
FIx notice should be success, not error.

ecgan committed:
Remove obsolete code comments.

ecgan committed:
Merge branch 'feature/refactor-shipping-rates-api' into feature/shipping-rates-api-ui-changes

ecgan committed:
Change export default for FlatShippingRatesInputCards.

ecgan committed:
Fix layout for the add rate button.

ecgan committed:
Code refactor with SHIPPING_RATE_METHOD.FLAT_RATE.

(cherry picked from commit 78dfd29fb7c9bf5dfe2f250ea0c33e998c4fa218)

ecgan committed:
Defensive check on shipping rate flat_rate method.

(cherry picked from commit 13f3255f60400417a2389ece8b110dcf331a9b50)

ecgan committed:
Change groupShippingRatesByPriceCurrency to also group by shipping rate method.

(cherry picked from commit 05968b4a592dab68f341ba74270c3d2bf9fc9f45)

ecgan committed:
Fix error with missing shipping rate method.

The error happens when we go to EditRateModal and add a country and save the changes.

(cherry picked from commit 34e7588e21a5b83313e8cca7ac23d094816871b2)

ecgan committed:
Merge branch 'feature/shipping-rates-ui/select-control-label' into feature/shipping-rates-ui/edit-free-listings

Conflicts:
	js/src/components/flat-shipping-rates-input-cards/estimated-shipping-rates-card/edit-rate-modal.js
	js/src/components/free-listings/configure-product-listings/shipping-rate/shipping-rate-setup/index.js
	js/src/data/actions.js
	js/src/data/reducer.js
	js/src/edit-free-campaign/index.js

ecgan committed:
Revert files in Setup MC, to have a better clean PR.

ecgan committed:
Make ShippingRateSection shared component and move FlatShippingRatesInputCards into it.

ecgan committed:
Rename ShippingRateSection countries prop to audienceCountries.

This is to make it consistent with child components.

ecgan committed:
CSS lint fix.

ecgan committed:
Better component import.

ecgan committed:
Rename groupShippingRatesByPriceCurrency to groupShippingRatesByMethodPriceCurrency.

eason9487 commented on js/src/components/shipping-rate-section/free-shipping-cards.js:
```diff
@@ -0,0 +1,58 @@
+	const { value, onChange } = props;
+
+	const initialOfferFreeShippingValue = value.some(
+		( el ) => el.options?.free_shipping_threshold > 0
```
❔ The possible types of `free_shipping_threshold` seem to be only `undefined` and number. Why does it need `?.` here?

---

And probably the default empty value of `options` given from API is an array. Even though `options` will be an object after being updated in frontend runtime, it looks buggy when checking `free_shipping_threshold` property to an array. Could we ask backend dev to help adjust it, or at least adapt it in actions.js?

![2022-03-01 16 36 03](https://user-images.githubusercontent.com/17420811/156134025-0c86da58-923d-47c2-ae44-e5f92da3b37f.png)

ecgan replied:
> The possible types of `free_shipping_threshold` seem to be only `undefined` and number. Why does it need `?.` here?

Fixed in [`562f38d` (#1266)](https://github.com/woocommerce/google-listings-and-ads/pull/1266/commits/562f38d7a10dcfdb344335f31a64a23512fa3ed5)

> Could we ask backend dev to help adjust it, or at least adapt it in actions.js?

@nima-karimi just came up with a change on that, see PR https://github.com/woocommerce/google-listings-and-ads/pull/1284.

eason9487 commented on js/src/components/shipping-rate-section/free-shipping-cards.js:
```diff
@@ -0,0 +1,58 @@
+const FreeShippingCards = ( props ) => {
+	const { value, onChange } = props;
+
+	const initialOfferFreeShippingValue = value.some(
+		( el ) => el.options?.free_shipping_threshold > 0
+	);
```
❓  From the visual design, the initial state is unchecked for any options. Has this requirement been changed?

![image](https://user-images.githubusercontent.com/17420811/156141135-01399b43-852e-42f2-8610-b9af336819bb.png)

ecgan replied:
The one in the visual design is based on Setup MC. When I work on this Edit Free Listings, I thought about this and realize it doesn't make sense to have this unchecked when people go to Edit Free Listings, so we have it checked based on whether they have the free shipping threshold or not.

eason9487 replied:
Sorry for my uncleared context. Yes, I have noticed the visual design is based on Setup MC. I meant here doesn't provide a way to let the initial state be unchecked for both options. After reviewing #1267, the initial state in Setup MC is indeed the same as here.

ecgan replied:
I gave some more thoughts about this and I have a better idea now. Instead of having the yes/no value depending on a state in the component, we can put it as a form value. Edit Free Listings can have an initial value based on the current code logic. Setup MC will have `undefined` initial value so users would have to yes/no there.

I created issue https://github.com/woocommerce/google-listings-and-ads/issues/1288 for this and I will make a separate PR on top of #1267 and make it work for Edit Free Listings and Setup MC.

eason9487 commented on js/src/utils/groupShippingRatesByMethodPriceCurrency.js:
```diff
@@ -33,41 +45,45 @@
+ * @param {Array<Object>} shippingRates Array of shipping rates in the format of `{ country, rate, currency }`.
  */
-const groupShippingRatesByPriceCurrency = ( shippingRates ) => {
+const groupShippingRatesByMethodPriceCurrency = ( shippingRates ) => {
```
💅  Two old names missed the renaming

https://github.com/woocommerce/google-listings-and-ads/blob/82bd0bc22cd14718ec9b01de0d23a23e025ff73d/js/src/utils/groupShippingRatesByMethodPriceCurrency.js#L42
https://github.com/woocommerce/google-listings-and-ads/blob/82bd0bc22cd14718ec9b01de0d23a23e025ff73d/js/src/utils/groupShippingRatesByMethodPriceCurrency.test.js#L6

ecgan replied:
Fixed in [`8ad41b3` (#1266)](https://github.com/woocommerce/google-listings-and-ads/pull/1266/commits/8ad41b3893e0abae4562678812fb8eca09175e97).

eason9487 commented on js/src/utils/groupShippingRatesByMethodPriceCurrency.js:
```diff
@@ -33,41 +45,45 @@
+		const { country, method, rate, currency } = shippingRate;
 		const price = Number( rate );
-		const priceCurrency = `${ price } ${ currency }`;
-		const group = rateGroupMap.get( priceCurrency ) || {
+		const methodPriceCurrency = `${ method } ${ price } ${ currency }`;
+		const group = rateGroupMap.get( methodPriceCurrency ) || {
 			countries: [],
+			method,
 			price,
```
📜 
In the codebase related to shipping rate processing, there are several places that use `price` as name of variables. Would this PR be an opportunity to change the variables of `price` used for shipping rate to `rate`, making #288 easier to move forward?

ecgan replied:
I thought about that and I had the itch of renaming that "price" term to match with "rate" term from API, but there are already a lot of things going on in this PR and in #1267, I feel it is better to leave it now and have it as a separate future PR after #1267. 😄

eason9487 commented on js/src/utils/groupShippingRatesByMethodPriceCurrency.js:
```diff
@@ -33,41 +45,45 @@
 	shippingRates.forEach( ( shippingRate ) => {
-		const { countryCode, rate, currency } = shippingRate;
+		const { country, method, rate, currency } = shippingRate;
 		const price = Number( rate );
```
💅  After #670, looks like we don't need to convert `rate` to number type anymore?

ecgan replied:
Fixed in [`13b5864` (#1266)](https://github.com/woocommerce/google-listings-and-ads/pull/1266/commits/13b5864f7b96aeed1c0004dd0b70d1054f5b2fb1).

eason9487 commented on js/src/components/shipping-rate-section/estimated-shipping-rates-card/estimated-shipping-rates-card.js:
```diff
@@ -73,58 +89,83 @@ export default function ShippingCountriesForm( {
+		countries.forEach( ( country ) => {
+			const oldShippingRate = actualCountries.get( country );
+			const newShippingrate = {
+				...defaultShippingRate,
```
💅  It seems like `defaultShippingRate` is not needed here?

ecgan replied:
I added that line in commit [`5b78fcc` (#1266)](https://github.com/woocommerce/google-listings-and-ads/pull/1266/commits/5b78fcc9bbba2b98a0c6c428002f283a4a550154). Without that line, an error would happen when we go to EditRateModal and add a country and save the changes. Also, I think it is also good to be defensive and make this `handleChange` consistent with `handleAdd`.

eason9487 replied:
Got it. Thanks!

eason9487 commented on js/src/components/shipping-rate-section/estimated-shipping-rates-card/shipping-rate-input-control.js:
```diff
@@ -71,25 +69,47 @@ const CountriesPriceInput = ( {
 				value={ price }
 				onBlur={ handleBlur }
 			/>
+			{ price === 0 && (
+				<div className="gla-input-pill-div">
+					<Pill>
+						{ __(
+							'Free shipping for all orders',
+							'google-listings-and-ads'
+						) }
+					</Pill>
+				</div>
+			) }
```
📜  or 💅 
We could consider moving the logic and pill into the \<EstimatedShippingRatesCard\> component as a sibling when rendering \<ShippingRateInputControl\>. For example:
```js
{ countriesPriceArray.map( ( el ) => {
    return (
        <div key={ el.countries.join( '-' ) }>
            <ShippingRateInputControl
                value={ el }
                audienceCountries={ audienceCountries }
                totalCountyCount={ totalCountyCount }
                onChange={ handleChange }
                onDelete={ handleDelete }
            />
            { el.price === 0 && (
                <div className="gla-input-pill-div">
                    <Pill>
                        { __(
                            'Free shipping for all orders',
                            'google-listings-and-ads'
                        ) }
                    </Pill>
                </div>
            ) }
        </div>
    );
} ) }
```

In this way, it would be easier to make \<ShippingRateInputControl\> to be a more generic component without a very similar \<MinimumOrderInputControl\>.

ecgan replied:
I intentionally put the Pill inside `ShippingRateInputControl`, with the idea of "when the shipping rate input control has 0 value, it will show a pill beside it". The component contains label, input and pill in it. 

> it would be easier to make `<ShippingRateInputControl>` to be a more generic component without a very similar `<MinimumOrderInputControl>`.

Another idea is to have something like `<GenericRateInputControl>`, that can be reused by `<ShippingRateInputControl>` and `<MinimumOrderInputControl>`. I intentionally want to separate ShippingRateInputControl and MinimumOrderInputControl because they do have slightly different behavior. 

We can address this later. 📜 😄

eason9487 reviewed (commented):
1. ❔ or 🚧  The form can continue to save even if the "Minimum order to qualify for free shipping" field is not filled in for any country. Is this the expected result?
    ![2022-03-01 19 13 12](https://user-images.githubusercontent.com/17420811/156159366-cebeb6d7-90b3-4135-a141-adbd7b1db4ea.png)
3. 🚧  The behavior of alerting user that there is unsaved data is not the same as before. Originally, even if user has edited it, the alert will not pop up if user changes it back to the original value.
   - The shipping rate

      https://user-images.githubusercontent.com/17420811/156161814-f3029a2b-b992-4cc7-a3ae-d15184d6a5ba.mp4
   - The free shipping threshold

      https://user-images.githubusercontent.com/17420811/156161880-84aa632f-cec4-47ab-a05a-719b1e7cfa1b.mp4
5. ❓ About some new components
   > - Partly addresses the tech debt: https://github.com/woocommerce/google-listings-and-ads/issues/288.

   Only finished partial code review, haven't looked closely at the parts in the **minimum-order-card** folder yet. However, it seems that this PR has added several new components that are more or less similar to the existing ones. For examples:
   - `<AddRateModal>` and `<AddMinimumOrderModal>`
   - `<EditRateModal>` and `<EditMinimumOrderModal>`
   - `<ShippingRateInputControl>` and `<MinimumOrderInputControl>`
   This makes me hesitate whether this PR might bring more things that need to be refactored in the future.

eason9487 commented on js/src/components/shipping-rate-section/minimum-order-card/minimum-order-card.js:
```diff
@@ -0,0 +1,174 @@
+		( shippingRate ) =>
+			shippingRate.rate > 0 &&
+			shippingRate.method === SHIPPING_RATE_METHOD.FLAT_RATE
+	);
```
💅 Looks like this code could be shared more or less with the following code.
https://github.com/woocommerce/google-listings-and-ads/blob/82bd0bc22cd14718ec9b01de0d23a23e025ff73d/js/src/components/shipping-rate-section/flat-shipping-rates-input-cards.js#L10-L14

ecgan replied:
Made a code refactor in [`b20ede6` (#1266)](https://github.com/woocommerce/google-listings-and-ads/pull/1266/commits/b20ede66be05419a5b24e815a540a77e73f7ed7a), with a new function called `isNonZeroFlatShippingRate` (not sure if you have a better naming? I don't. 😂 )

eason9487 replied:
> new function called `isNonZeroFlatShippingRate` (not sure if you have a better naming? I don't. 😂 )

No idea either. 😂 

And I'm slightly more leaning towards the below approach:

```js
// util
function getChargedShippingRates( rates, method = SHIPPING_RATE_METHOD.FLAT_RATE ) {
    return rates.filter(
        ( shippingRate ) =>
            shippingRate.rate > 0 && shippingRate.method === method
        );
}

// <MinimumOrderCard>
const nonZeroShippingRates = getChargedShippingRates( value );

// <FlatShippingRatesInputCards>
const displayFreeShippingCards = getChargedShippingRates( value ).length > 0;
```
Even though all countries were selected, there should be no significant difference in the calculation performance here.

No matter which way, it's already better. Thanks for the adjustment!

ecgan replied:
> ```js
> getChargedShippingRates
> ```

You gave me a better naming idea, I have changed the function name from `isNonZeroFlatShippingRate` to `isNonFreeFlatShippingRate` (also based on wiktionary page on "[nonfree](https://en.wiktionary.org/wiki/nonfree)"), in [`85e0033` (#1266)](https://github.com/woocommerce/google-listings-and-ads/pull/1266/commits/85e00336bcba72cc8b3e7c5a51fada4e030c00b5). 😄 

> And I'm slightly more leaning towards the below approach

I still prefer the current `isNonFreeFlatShippingRate` way, because the function is smaller, more reusable and more readable, in my opinion. 💅 😄 🙏 

```js
// <MinimumOrderCard>
const nonZeroShippingRates = value.filter( isNonFreeFlatShippingRate );

// <FlatShippingRatesInputCards>:
const displayFreeShippingCards = value.some( isNonFreeFlatShippingRate );

// And we can use it if we are dealing with one shipping rate.
const isNonFreeFlat = isNonFreeFlatShippingRate( shippingRate )
```

eason9487 commented on js/src/components/shipping-rate-section/minimum-order-card/groupShippingRatesByFreeShippingThreshold.js:
```diff
@@ -0,0 +1,24 @@
+
+	shippingRates.forEach( ( shippingRate ) => {
+		const threshold =
+			shippingRate.options?.free_shipping_threshold === undefined
```
❔ The possible types of `free_shipping_threshold` seem to be only `undefined` and number. Why does it need `?.` here?

ecgan replied:
Fixed in [`4871ff5` (#1266)](https://github.com/woocommerce/google-listings-and-ads/pull/1266/commits/4871ff54f79760730d63c241276e1e90adb50af0).

eason9487 commented on js/src/components/shipping-rate-section/minimum-order-card/groupShippingRatesByFreeShippingThreshold.js:
```diff
@@ -0,0 +1,24 @@
+		const threshold =
+			shippingRate.options?.free_shipping_threshold === undefined
+				? undefined
+				: Number( shippingRate.options.free_shipping_threshold );
```
💅 If the previously stored `free_shipping_threshold` value fetched from API is always number type, the typecasting here could be omitted.

ecgan replied:
Fixed in [`4871ff5` (#1266)](https://github.com/woocommerce/google-listings-and-ads/pull/1266/commits/4871ff54f79760730d63c241276e1e90adb50af0).

eason9487 commented on js/src/components/shipping-rate-section/minimum-order-card/minimum-order-card.js:
```diff
@@ -0,0 +1,174 @@
+							modal={
+								<AddMinimumOrderModal
+									countryOptions={ countryOptions }
+									value={ emptyThresholdGroup }
```
🚧 or❓
When there is more than one country to be set free shipping threshold, countries that already have threshold value will still appear in the country selector as options. It is inconsistent with the UI behavior of shipping rates setup. Is this the expected result?

https://user-images.githubusercontent.com/17420811/156349827-4b22a8ff-6d78-44c3-b772-24c425be5408.mp4

ecgan replied:
> It is inconsistent with the UI behavior of shipping rates setup.

Shipping rates setup behave the same way for me, I can always manually select a country in the dropdown even though it has already been setup in another row...? 

https://user-images.githubusercontent.com/417342/156592105-9b623f7c-e72c-4e77-ac45-cb3865802d4d.mov

eason9487 replied:
The inconsistency is between `<AddMinimumOrderModal>` and `<AddRateModal>`, not between `<EditRateModal>` and others.

ecgan replied:
Gotcha! I have made it consistent with AddRateModal, in [`943ce4c` (#1266)](https://github.com/woocommerce/google-listings-and-ads/pull/1266/commits/943ce4c7500168e957e1651442262784fd9d4a42).

Side note: comparing Add modal and Edit modal, I do feel like it is inconsistent, I feel like we should make Add modal behave like Edit modal and make all the countries show up, because if all the countries can show up in Edit modal, then why don't they show up in Add modal? 🤔 For future iterations. 📜 😄

eason9487 replied:
> Side note: comparing Add modal and Edit modal, I do feel like it is inconsistent, I feel like we should make Add modal behave like Edit modal and make all the countries show up, because if all the countries can show up in Edit modal, then why don't they show up in Add modal?

Agree, I feel a bit concerned about the inconsistency. I didn't notice Add modal and Edit modal are using different country options until viewing this PR. 😂  And I couldn't recall if there's any specific purpose. So I tried to deduce the possibilities.

This is what I guessed when reading the codes:
- When editing, listing all countries make merchants easier to move countries from other groups into the current editing one, which reduces the many UI operations.
- When adding, listing remaining countries let merchants easier to see which countries are not yet set.

> For future iterations. 📜 😄

Regardless using all countries or remaining countries on which modal, I guess they both are reasonable. I also think we could go back later to see if it needs adjustment.

eason9487 commented on js/src/components/shipping-rate-section/minimum-order-card/minimum-order-card.js:
```diff
@@ -0,0 +1,174 @@
+		( shippingRate ) => shippingRate.country
+	);
+
+	const handleEditChange = ( oldGroup ) => ( newGroup ) => {
```
🚧 
When entering `0` from different input controllers, the value of passed-in arguments `new Group.threshold` is inconsistent, resulting in different results.

https://user-images.githubusercontent.com/17420811/156354408-e8c2f79c-3893-4415-a01a-b57e869012dd.mp4

ecgan replied:
Fixed in [`af7f16a` (#1266)](https://github.com/woocommerce/google-listings-and-ads/pull/1266/commits/af7f16a8ec39c6f6998966d54f85a282d833a492).

eason9487 commented on js/src/components/shipping-rate-section/minimum-order-card/add-minimum-order-modal.js:
```diff
@@ -0,0 +1,110 @@
+			initialValues={ {
+				countries: value.countries,
+				threshold: value.threshold,
+				currency: value.currency,
```
💅  Looks like the `currency` here is unneeded?

ecgan replied:
I want the currency to go side-by-side together with the threshold value. I think that is a better design in the long term, in case there are scenarios like different thresholds with different currencies, this would be able to work with that. Also, previously there was a related issue https://github.com/woocommerce/google-listings-and-ads/issues/1189 in shipping rate, where we were only looking at number without currency. I'd like to avoid that kind of situation.

Currency is actually used as `values.currency` in `AppInputPriceControl`'s `suffix` props.

eason9487 replied:
❔ 

> I think that is a better design in the long term, in case there are scenarios like different thresholds with different currencies, this would be able to work with that.

If the currency is different, there is still no need to put it in `initialValues`. If consider that the currency could be edited in the future, then it would be more obvious at that time whether to put it in `initialValues` or to use another way to handle it when the actual need appears.

When reading codes, it is a bit confusing to see that `currency` is not a form item and does not need to be passed down to any separated sub-components, but only to be passed to the `values` below and then placed in the `suffix` prop.

> Also, previously there was a related issue #1189 in shipping rate, where we were only looking at number without currency. I'd like to avoid that kind of situation.

I couldn't understand why dose placing `currency` in `initialValues` relate to issue #1189? 🤔 

> Currency is actually used as `values.currency` in `AppInputPriceControl`'s `suffix` props.

It could be changed to `suffix={ value.currency }`.

ecgan replied:
> When reading codes, it is a bit confusing to see that `currency` is not a form item and does not need to be passed down to any separated sub-components...

I guess your understanding or expectation is that when we see the Form `initialValues`, the values should have a corresponding UI that can be seen and/or modified by the users, and in code level, there should be a corresponding `{ ...getInputProps( 'fieldname' ) }` for the value?

My understanding on Form `initialValues` is more than that though. I see it as a logical grouping of things. In some cases, it can be used as a Form "state" without needing to use `useState`. The values may or may not have a corresponding UI or a corresponding `{ ...getInputProps( 'fieldname' ) }`. Sometimes a value may be a derived value based on other values (e.g. using `setValue`), and we want to group them together. Sometimes a value may participate in validation.

I gave some thoughts about your comment: `currency` is not modifiable by users, there is no `setValue` for it, and it does not participate in validation. I proceeded to remove `currency` from `initialValues`, and when I was trying it out, I realized that because `currency` is no longer grouped together, I had to make a bit modification to `handleSubmitCallback` to achieve the same outcome:

```js
// (Existing) Approach 1: newValue has countries, threshold and currency.
const handleSubmitCallback = ( newValue ) => {
	onRequestClose();
	onChange( newValue );
};

// (Suggested) Approach 2. newValue has countries and threshold, no currency.
const handleSubmitCallback = ( newValue ) => {
	onRequestClose();
	onChange( {
		... newValue,
		currency: value.currency
	} );
};
```

When I look at the above, I prefer Approach 1. It relates to my point on logical grouping of things. The shape of `value` is consistent throughout and requires less mental process, in my opinion.

Keen to hear your opinion on this. 😄  

> ... but only to be passed to the `values` below and then placed in the `suffix` prop.

> It could be changed to `suffix={ value.currency }`.

Because `currency` _is already in the form values_, the child should then get the value from the Form `values` instead of the outer `value` in the AddMinimumOrderModal component. This would make things easier in case we want to do refactoring on the child code.

> > Also, previously there was a related issue #1189 in shipping rate, where we were only looking at number without currency. I'd like to avoid that kind of situation.
> 
> I couldn't understand why dose placing `currency` in `initialValues` relate to issue #1189? 🤔

They are _not technically code related_. What I meant to say is the _idea_ to look at _the price (the number and the currency) as a logical group_. (Sorry 😄 🙏 )

eason9487 replied:
Thanks for the explanation!

> I guess your understanding or expectation is that when we see the Form `initialValues`, the values should have a corresponding UI that can be seen and/or modified by the users, and in code level, there should be a corresponding `{ ...getInputProps( 'fieldname' ) }` for the value?

> My understanding on Form `initialValues` is more than that though. I see it as a logical grouping of things.

From my understanding, the `initialValues` is more likely the values that are used closely with form and its fields. Even if there are other logically related (grouping) values, adding them might still be relatively not ideal if there are no special situations.

Because the event and state handling of forms are often relatively complex, using it as a logical grouping of things might increase the complexity.

> In some cases, it can be used as a Form "state" without needing to use `useState`.

Sorry I'm not quite sure whether borrowing a \<Form\> value as a state is a good idea. It will trigger \<Form\>'s `onChange` callback when calling `setValue`, which makes callback to be triggered when there is no "real" form values changed.

And that state will also appear in the arguments of `validate` and `onSubmit` callbacks. The mix of form values and states in \<Form\> callbacks could be confusing when reading the code, further reducing readability and complexity when maintaining it.

> The values may or may not have a corresponding UI or a corresponding `{ ...getInputProps( 'fieldname' ) }`.

It's useful when values have corresponding `{ ...getInputProps( 'fieldname' ) }`. But sorry I'm not so sure why it's useful when the values may or may not have a corresponding UI? Seems they are more like props.

> Sometimes a value may be a derived value based on other values.

If there is a derived value based on form values, It could be handled in \<Form\> `onChange`. In this way, it provides a much clear overview to figure out how this derived value is updated accordingly, and lower coupling.

> Sometimes a value may participate in validation.

This is indeed more convenient, but I believe if a logical grouping value that participates in validation could be put in `initialValues`, then usually it was already accessible in `validate` callback as well. So, when their relationship is in validation scope, would it be a bit better to make them logically grouped in the `validate` callback rather than the `initialValues`? 

---

> I proceeded to remove `currency` from `initialValues`, and when I was trying it out, I realized that because `currency` is no longer grouped together, I had to make a bit modification to `handleSubmitCallback` to achieve the same outcome:
> ```js
> // ... omitted
> // (Suggested) Approach 2. newValue has countries and threshold, no currency.
> const handleSubmitCallback = ( newValue ) => {
> 	onRequestClose();
> 	onChange( {
> 		... newValue,
> 		currency: value.currency
> 	} );
> };
>
> ```
> When I look at the above, I prefer Approach 1. It relates to my point on logical grouping of things. The shape of value is consistent throughout and requires less mental process, in my opinion.
> 
> Keen to hear your opinion on this.

Sorry, I'm not quite sure why does the same shape of `value` here can requires less mental process nor how to evaluate when things have logical grouping relation. In my experience, adding JSDocs for them might be more intuitive and effective in reducing mental processes.

And I think if removing `currency` was ever a choice to be used, then maintaining a consistent shape should no longer be a need since they are already inconsistent in the `initialValues`. And because `currency` is not on any of above examples listed for logical grouping, then maintaining logical grouping should not be a consideration either.

I commented whether `currency` could be removed at a few places in this PR because I went back and forth several times between these files during the first code review, to see where `currency` was being used. To think about whether there are any edge cases where `currency` is needed that I didn't notice. And this process was repeated twice when I replied to this comment.

After repeatedly verifying that `currency` just needs to be passed to `suffix`, I thought deleting them could improve readability and maintainability. Otherwise, future maintenance might still cause similar back-and-forth and validation processes as above.

> Because `currency` _is already in the form values_, the child should then get the value from the Form `values` instead of the outer `value` in the AddMinimumOrderModal component. This would make things easier in case we want to do refactoring on the child code.

Sorry, I'm not sure I agree that this is a "should" convention/principle because it seems a bit of reverse causation, and the `value.currency` could even be a `suffix` prop of \<AddMinimumOrderModal\> rather than mixing in `value` prop. Based on above comments, binding it with `value` prop makes it be confusing, in my opinion.

And if how easier when refactoring is a concern at this point, then it could be handled in the same way as `countryOptions` and `onRequestClose` when it happens.

---

After the above discussion, I knew that there are some reasons for keeping the same shape/logical grouping. Although I have some concerns about readability and maintainability, that is just my subjective opinion.

Given that the implementation is workable, and that my comment started with a 💅 . I think this PR can be moved on regardless of which option is chosen.

ecgan replied:
Thanks so much for the explanation! 

> I commented whether `currency` could be removed at a few places in this PR because I went back and forth several times between these files during the first code review, to see where `currency` was being used. To think about whether there are any edge cases where `currency` is needed that I didn't notice. And this process was repeated twice when I replied to this comment.

I think this one above is especially noteworthy for me. Now I know your perspective better. I got to let it sink in my mind for a bit and see how it turns out. 😄

eason9487 commented on js/src/components/shipping-rate-section/minimum-order-card/edit-minimum-order-modal.js:
```diff
@@ -0,0 +1,126 @@
+		<Form
+			initialValues={ {
+				countries: value.countries,
+				currency: value.currency,
```
💅  Looks like the `currency` here is unneeded?

ecgan replied:
Same as https://github.com/woocommerce/google-listings-and-ads/pull/1266#discussion_r818990917.

eason9487 commented on js/src/components/shipping-rate-section/minimum-order-card/edit-minimum-order-modal.js:
```diff
@@ -0,0 +1,126 @@
+		onRequestClose();
+		onChange( {
+			countries: [],
+			currency: undefined,
```
❓ Why here needs to pass `currency` as `undefined`?

ecgan replied:
I just want to make the object have the same consistent schema as the form value object `{ countries, currency, threshold }`.

eason9487 replied:
Got it.

But if consider removing `currency` from `initialValues` due to https://github.com/woocommerce/google-listings-and-ads/pull/1266#discussion_r817609571, which is a similar concern mentioned in https://github.com/woocommerce/google-listings-and-ads/pull/1266#discussion_r819353999, then here could be removed, too.

And if aim to keep it, I just feel resetting `currency` to `undefined` looks buggy since currency should never be `undefined` during shipping setup.

ecgan replied:
> I just want to make the object have the same consistent schema as the form value object `{ countries, currency, threshold }`.

Sorry, there was a slight mistake I said above. I actually meant to say I want to make the object have the same consistent schema as the `value` props. The idea is that the consumers pass a value into the component via `value` prop, and they get the new value from the component via `onChange` prop. I'd like the old value and the new value to have the same consistent schema.

> I just feel resetting `currency` to `undefined` looks buggy since currency should never be `undefined` during shipping setup.

I was thinking since this is a delete operation, it would make sense to have `currency: undefined` like `threshold: undefined`, i.e. everything is deleted.

eason9487 commented on js/src/utils/groupShippingRatesByMethodPriceCurrency.js:
```diff
@@ -6,24 +6,36 @@
+ *         method: "flat_rate",
  *         currency: 'USD',
  *         rate: 20,
+ *         options: [],
```
Note: Related to the initial values mentioned in https://github.com/woocommerce/google-listings-and-ads/pull/1266#discussion_r816551941

ecgan replied:
Updated in [`12e66c0` (#1266)](https://github.com/woocommerce/google-listings-and-ads/pull/1266/commits/12e66c00f228666a2445398a867f2753020c0b78) in accordance with PR #1284.

eason9487 commented on js/src/utils/groupShippingRatesByMethodPriceCurrency.test.js:
```diff
@@ -1,48 +1,63 @@
+				method: 'flat_rate',
 				currency: 'USD',
 				rate: 20,
+				options: [],
```
Note: Related to the initial values mentioned in https://github.com/woocommerce/google-listings-and-ads/pull/1266#discussion_r816551941

ecgan replied:
Updated in [`12e66c0` (#1266)](https://github.com/woocommerce/google-listings-and-ads/pull/1266/commits/12e66c00f228666a2445398a867f2753020c0b78) in accordance with PR #1284.

eason9487 commented on js/src/components/shipping-rate-section/minimum-order-card/minimum-order-input-control.js:
```diff
@@ -0,0 +1,68 @@
+		onChange( {
+			countries,
+			threshold: numberValue === 0 ? undefined : numberValue,
+			currency,
```
💅 Looks like the `currency` here is unneeded?

ecgan replied:
Same as https://github.com/woocommerce/google-listings-and-ads/pull/1266#discussion_r818990917.

eason9487 replied:
Similar to the "But if consider removing [...]" section in https://github.com/woocommerce/google-listings-and-ads/pull/1266#discussion_r819361044

eason9487 commented on js/src/components/shipping-rate-section/minimum-order-card/minimum-order-input-control.js:
```diff
@@ -0,0 +1,68 @@
+
+		onChange( {
+			countries,
+			threshold: numberValue === 0 ? undefined : numberValue,
```
💅  or 🚧 
If `0` is an invalid value, I'd suggest moving this logic to `<MinimumOrderCard>` and fix the inconsistency logics at:
1. https://github.com/woocommerce/google-listings-and-ads/blob/82bd0bc22cd14718ec9b01de0d23a23e025ff73d/js/src/components/shipping-rate-section/minimum-order-card/add-minimum-order-modal.js#L30-L35
2. https://github.com/woocommerce/google-listings-and-ads/blob/82bd0bc22cd14718ec9b01de0d23a23e025ff73d/js/src/components/shipping-rate-section/minimum-order-card/edit-minimum-order-modal.js#L38-L43

ecgan replied:
Fixed in [`af7f16a` (#1266)](https://github.com/woocommerce/google-listings-and-ads/pull/1266/commits/af7f16a8ec39c6f6998966d54f85a282d833a492).

eason9487 commented on js/src/components/shipping-rate-section/minimum-order-card/minimum-order-input-label-text.js:
```diff
@@ -0,0 +1,43 @@
+	return (
+		<div>
+			{ createInterpolateElement( string, {
+				countries: <strong>{ firstCountryNames.join( ', ' ) }</strong>,
+				count: <>{ countries.length - firstCountryNames.length }</>,
```
💅  It would be better to use `sprintf` for dynamic values and only uses `createInterpolateElement` for wrapping HTML tags or react elements if it's possible.

ecgan replied:
I thought about `sprintf`, but the thing is we want to bold the country names in the middle of the string, so `sprintf` can't help us with that?

eason9487 replied:
It could be done in this way:
https://github.com/woocommerce/google-listings-and-ads/blob/c492680956b0d0e198a4512a87527a41ea5e9134/js/src/components/paid-ads/budget-section/budget-recommendation/index.js#L32-L50

ecgan replied:
Thanks for the example! Updated in [`4e9cfe9` (#1266)](https://github.com/woocommerce/google-listings-and-ads/pull/1266/commits/4e9cfe97b38524bd35f2146b5d0d2f6e99dd37ea), and it has a much better HTML output. 😃 👍

ecgan committed:
Merge branch 'feature/shipping-rates-ui/select-control-label' into feature/shipping-rates-ui/edit-free-listings

ecgan committed:
Remove unneeded optional chaining in FreeShippingCards.

ecgan committed:
Fix missed renaming for groupShippingRatesByMethodPriceCurrency.

ecgan committed:
Remove unneeded Number() conversion in groupShippingRatesByMethodPriceCurrency.

ecgan committed:
Code refactor with isNonZeroFlatShippingRate.

ecgan committed:
Simplify groupShippingRatesByFreeShippingThreshold.

We don't need optional chaining, and free_shipping_threshold is always number from the API.

ecgan committed:
Better validation for minimum order input.

Check for threshold must be greater than 0.

ecgan committed:
Merge branch 'feature/shipping-rates-ui/main' into feature/shipping-rates-ui/edit-free-listings

ecgan committed:
Rename isNonZeroFlatShippingRate to isNonFreeFlatShippingRate.

ecgan committed:
Make AddMinimumOrderModal display options using countries with no threshold value.

This is to match with AddRateModal behavior.

ecgan committed:
Better HTML output with sprintf in MinimumOrderInputLabelText.

ecgan committed:
Change options from [] to {}.

This is in accordance to https://github.com/woocommerce/google-listings-and-ads/pull/1284.

ecgan commented:
@eason9487 ,

> 1. ❔ or 🚧  The form can continue to save even if the "Minimum order to qualify for free shipping" field is not filled in for any country. Is this the expected result?

I've created issue https://github.com/woocommerce/google-listings-and-ads/issues/1289 for this.

> 2. 🚧  The behavior of alerting user that there is unsaved data is not the same as before. Originally, even if user has edited it, the alert will not pop up if user changes it back to the original value.

The users change the form values and the form is considered "dirty", so I think it is probably still fine for us to prompt the users on this. WDYT?

Nevertheless, I have created issue https://github.com/woocommerce/google-listings-and-ads/issues/1290 for this if we want to match back with existing behavior. I think it is at a lower priority than other issues we found in this PR though.

> However, it seems that this PR has added several new components that are more or less similar to the existing ones. For examples:
> 
> * `<AddRateModal>` and `<AddMinimumOrderModal>`
> * `<EditRateModal>` and `<EditMinimumOrderModal>`
> * `<ShippingRateInputControl>` and `<MinimumOrderInputControl>`
>   This makes me hesitate whether this PR might bring more things that need to be refactored in the future.

I intentionally want to split "Rate" and "Minimum order", because I see them as different domain objects and they might have different logics in it. 

For example, in WooCommerce Free Shipping settings, users are able to specify "minimum order amount and a coupon" (see image below). Even though we don't have the coupon for free shipping feature now, the idea I like to convey is that "Rate" and "Minimum order" are different.

<img width="881" alt="image" src="https://user-images.githubusercontent.com/417342/156926408-f424dfd5-aca2-45cd-9d70-711dd0bbe064.png">

After addressing some of your review comments above, as of now, they have slightly different code behavior, e.g. different comparison check on rate and threshold in the Add modals.

I also noticed the code similarities, and thought about refactoring them. If we want to do that, we can refactor out similar code into a shared component, and have both Rate and MinimumOrder components consume the shared component. I didn't do that because I think things might still be changing and the value is probably not worth the effort. Maybe we can look into that later.

eason9487 commented:
> The users change the form values and the form is considered "dirty", so I think it is probably still fine for us to prompt the users on this. WDYT?
> 
> Nevertheless, I have created issue #1290 for this if we want to match back with existing behavior. I think it is at a lower priority than other issues we found in this PR though.

I imagined that as a user, which one is much closer to its meaning when I see a prompt message - "You have unsaved data. Are you sure you want to leave?":
1. There is exactly an unsaved change. Am I sure I don't want to save?
2. Editing has happened, it doesn't matter whether the final data is the same as the previously saved, I need to save the data. Otherwise, something will be lost.
3. Once I have touched the form, regardless of the value in the form field is the same as before when I moved out the input cursor, there is already something that needs to be saved before leaving.
4. Am I missing something?

If I'm not sure whether I edited data, the first one is a very intuitive understanding way. If I'm pretty sure the edited data is the same as the previously saved, then 2. 3. and 4. are the possibilities I can think of. But I can't confirm whether leaving will be a problem or not, so I have to save it.

How to implement the concept of "form is considered dirty" to users so that they can intuitively understand what is happening?

Moreover, the current inconsistencies go beyond the two examples listed. If there is a free shipping threshold in the initial form values, the prompt will not appear when the edited data is the same. At this point, its appearance timing is not due to dirty form again.

https://user-images.githubusercontent.com/17420811/156969616-e9282c3e-d6e7-4ca9-999f-9a6c82c42f07.mp4

eason9487 reviewed (approved):
Thanks for resolving most of the concerned comments! The remaining parts are not blocked, or they have been created follow-up issues. This PR is ready to move on.

ecgan commented:
> I imagined that as a user, which one is much closer to its meaning when I see a prompt message - "You have unsaved data. Are you sure you want to leave?":

> Moreover, the current inconsistencies go beyond the two examples listed. If there is a free shipping threshold in the initial form values, the prompt will not appear when the edited data is the same. At this point, its appearance timing is not due to dirty form again.

Thanks for the thoughtful explanation, and detailed testing with videos! 👍 I'll look into that in #1290 .
